web_src/css/themes/theme-gitea-light-protanopia-deuteranopia.css

```diff
@@ -0,0 +1,11 @@
+@import "./theme-gitea-light.css";
+
+/* red/green colorblind-friendly colors */
+/* from GitHub: --diffBlob-addition-*, --diffBlob-deletion-*, etc */
+:root {
+  --color-diff-added-word-bg: #54aeff66;
+  --color-diff-added-row-bg: #ddf4ff80;
+
+  --color-diff-removed-word-bg: #ffb77c80;
+  --color-diff-removed-row-bg: #fff1e580;
+}
```

Comment:
Do we really need more than one colorblind theme? These names are awefully long. I prefer to just name them "colorblind".

Reply:
Problem: There are three different types of colorblindness, and that's their name…

Reply:
Can they actually be combined like is done currently? Other UI's that I've seen have 3 separate options, GitHub on the other hand only has one option.

Reply:
I copied the names from GitHub, GitHub has much more types:

"protanopia deuteranopia" means "red/green" color blindness, "tritanopia" means "blue" color blindness (my dictionary tells me).

I also tried some names like: `gitea-light-red-green` or `gitea-light-red-green-friendly` or `gitea-light-colorblind-red-green` or `gitea-light-colorblind-red-green-friendly`, none of them seems better.

![image](https://github.com/go-gitea/gitea/assets/2114189/49f95903-70ff-402b-b998-55a5c2248231)

Reply:
Maybe shorten to `light-colorblind-pd`.

Reply:
What is `pd`? I will always say no to the improper abbreviations. Please do not introduce anything like `df ac fw f1` any more.

Reply:
pd = "protanopia deuteranopia", and it has nothing to do with those CSS classes.

Reply:
Well, it would cause confusion to end users. I guess it's impossible to an end user to understand pd = "protanopia deuteranopia".

I also agree that current approach is far from ideal, it only adds "initial support" (and fix some existing bugs like non-existing theme would make users see broken UI).

To improve, in my mind there could be somethings to do:

1. Introduce some better names than use the CSS name directly (the problem is where&how to store the names)
2. Add some theme previews, then end users could know what a theme would look like
3. Use some dedicated CSS classes for some elements (just like positive/negative, addition/deletion, etc), but not just use red/green names.

Reply:
Yes, themes need a "display name". I already proposed a mechanism for that with JSON in the themes config field in https://github.com/go-gitea/gitea/pull/27576.

Reply:
> Add some theme previews, then end users could know what a theme would look like

Quite hard, but might be possible with iframe. Still some places like footer would not have place for such previews.

> Use some dedicated CSS classes for some elements (just like positive/negative, addition/deletion, etc), but not just use red/green names.

I recommend primary,success,warning,danger.

Reply:
> Yes, themes need a "display name". I already proposed a mechanism for that with JSON in the themes config field in #27576.

My idea is to use some encoding rules for the filename. For example:

* `gitea-auto` => `Gitea Auto` (or trim all `gitea-` prefix and only get `Auto`)
* `My~20A~2fB` => decode as `My%20A%2fB` and get `My A/B`, not using `%` because some users don't configure the reverse proxy correctly, the `%xx` would be decoded incorrectly ahead by the reverse proxy.

Another idea is to store the display name into the CSS file (eg: by a `var`), and do a quick parse from each CSS file.




> > Add some theme previews, then end users could know what a theme would look like
> 
> Quite hard, but might be possible with iframe. Still some places like footer would not have place for such previews.

It could only show some key elements, like background, primary button, addition, deletion.

Reply:
Theme name from a CSS variable might be nice and I also had another use case where I'd need the `--is-dark-theme` in template context. Maybe https://github.com/tdewolff/parse can do it. Otherwise yeah, we can come of with a filename to display name function.

> It could only show some key elements, like background, primary button, addition, deletion.

If we already have the CSS parser in backend, it should be easy to extract a few variables for the preview, yes.

Reply:
> > It could only show some key elements, like background, primary button, addition, deletion.
> 
> If we already have the CSS parser in backend, it should be easy to extract a few variables for the preview, yes.

I think we could make it easier without a parser, just by a iframe, and we only generate a simple demo page like this:

```
<style {the-chosen-theme-css-file}>
<body>
 demo button
 demo addition/deletion
</body>
```

Reply:
> Theme name from a CSS variable might be nice and I also had another use case where I'd need the `--is-dark-theme` in template context. Maybe https://github.com/tdewolff/parse can do it. Otherwise yeah, we can come of with a filename to display name function.

IMO a regexp is good enough, we don't need to really introduce a heavy CSS parser .......

Reply:
> I think we could make it easier without a parser, just by a iframe

iframe is too slow to load in general, maybe there is a way to scope a css file to a certain element.

Reply:
Looks like it's possible to avoid iframe with shadow dom:

https://developer.mozilla.org/en-US/docs/Web/API/Web_components/Using_shadow_DOM#adding_style_elements_in_template_declarations